packages/catalog-server/src/lib/firestore/package-info-converter.ts

```diff
@@ -9,33 +9,67 @@ import {
   FirestoreDataConverter,
   QueryDocumentSnapshot,
   Timestamp,
+  WithFieldValue,
 } from '@google-cloud/firestore';
 
 import {
   DistTag,
+  isReadablePackage,
   PackageInfo,
   ReadablePackageInfo,
 } from '@webcomponents/catalog-api/lib/schema.js';
 
-export const packageInfoConverter: FirestoreDataConverter<
-  Omit<PackageInfo, 'version'>
-> = {
-  fromFirestore(
-    snapshot: QueryDocumentSnapshot<DocumentData>
-  ): Omit<PackageInfo, 'version'> {
+export const packageInfoConverter: FirestoreDataConverter<PackageInfo> = {
+  fromFirestore(snapshot: QueryDocumentSnapshot<DocumentData>): PackageInfo {
     const distTags = snapshot.get('distTags');
     const graphQLDistTags = Object.entries(distTags).map(
       ([tag, version]) => ({tag, version} as DistTag)
     );
     return {
-      name: snapshot.id,
+      name: idToPackageName(snapshot.id),
       lastUpdate: (snapshot.get('lastUpdate') as Timestamp).toDate(),
       status: snapshot.get('status'),
       description: snapshot.get('description'),
       distTags: graphQLDistTags,
+      // `version` is left to a sub-collection query
     } as ReadablePackageInfo;
   },
-  toFirestore(_packageInfo: PackageInfo) {
-    throw new Error('not implemented');
+  toFirestore(packageInfo: WithFieldValue<PackageInfo>) {
+    if (isReadablePackage(packageInfo as PackageInfo)) {
+      const data = packageInfo as WithFieldValue<ReadablePackageInfo>;
+      return {
+        status: data.status,
+        // TODO (justinfagnani): we could force this to be
+        // FieldValue.serverTimestamp() here.
+        lastUpdate: data.lastUpdate,
+        description: data.description,
+        distTags: new Map(
+          // We don't support FieldValues in distTags, so cast away:
+          (data.distTags as DistTag[]).map((t) => [t.tag, t.version])
+        ),
+      };
+    } else {
+      return {
+        status: packageInfo.status,
+        lastUpdate: packageInfo.lastUpdate,
+      };
+    }
   },
 };
+
+/**
+ * Converts a package name to a Firestore ID.
+ *
+ * Firestore IDs cannot include a '/', so we convert it to '__'.
+ *
+ * This is a similar to the transform TypeScript does for `@types`
+ * packages.
+ */
+export const packageNameToId = (packageName: string) =>
+  packageName.replace('/', '__');
```

Comment:
Just to confirm, `replace` is fine (vs `replaceAll`) because there can only be one "/" in a package name?

Reply:
Exactly. This is also the same transform that TypeScript uses in the `@types` packages for scoped sub-packages.